### ISSUE TITLE
Board index CSS: suggestion.

### ISSUE THREAD

**Issue by StealthWombat** [SimpleMachines/SMF#2665]
This is not really a bug as such, but is worth considering.

You have the board index rows marked up as four divs, with the CSS set to float them, and I noticed that you have set the sum of the widths to a bit less than 100%. I know why this is too: because if it's set to 100% the horizontal paddings on some of  the divs will blow the layout to pieces.

This is ok, but it's brittle. It won't play well with changes in screen width, since the divs will get smaller but the padding won't, so the total width of each div in % terms will increase, so it will probably require extra CSS to keep it under control. It will also be prone to blowing apart if someone changes paddings or adds left or right borders, meaning it wont be noob-friendly for themers and admins.

There are two ways around this. Both are easy. The first is to declare box-sizing: border-box; on the four divs. This will make their overall widths constant regardless of horizontal padding or side borders. You can the set them all to add up to exactly 100%, which will make sense for beginners.

The second option, which is equally easy, is to declare display: table-cell; on the divs. This will make them nice and rubbery like real table cells, even if the widths add up to more than 100%. As long as the parent for the rows is set to display: table; and table-layout: fixed; with the rows themselves set to display: table-row; and the four divs inside the rows set to table-cell, the whole thing will be very stable with a range of content and screen sizes.

Another advantage here is built-in RTL support, since it still behaves like a table if an RTL language is loaded. The order of the cell divs will automatically switch to match RTL, so you will need less code in rtl.css.

Obviously display can still be changed on small screens to stack content down the page.

Food for thought. :)

**Comment by XinYenFon** [comment]
I was going to apply the changes here but, point is I'm using reusable basic codes theory here :P (it looks cool when I say that way). Its not 100% because icon has min-width: 60px; also same code used in topic listing messageindex.

**Comment by StealthWombat** [comment]
Basic re-usable code theory my arse. :P 

Just noticed this is still in Beta 2. If you declare as table-cell it will be a lot more stable, and you can still re-use it anywhere you like. It will also give automatic RTL support, so you will be able to ditch some code from rtl.css.

Basically, using display: table-cell; is ideal anywhere you want to convince standard block level elements to behave like table cells (all in a row, and nice and tolerant of content and styling changes).

The way it is in Beta 2 is going to be less n00b-friendly. It won't bother experienced themers. They'll just write their own CSS to declare as table-cell or border-box or flying-kittens-in-fishnet-stockings or whatever. It will bother the n00bs though, because it will make the theme more breakable and they won't know why.

Suit yourself if you want to aggravate n00bz.

**Comment by live627** [comment]
I like that idea.

Also, how would it work with blocks hon the  stats page?

I really would like to use flexboxes for these things but IE9.

**Comment by Rupurudu** [comment]
First, let's talk about why someone decided to ditch the original design. Why we are not using real tables anymore?

**Comment by XinYenFon** [comment]
@StealthWombat agree to disagree. Since you don't have to rebuild the template I think its better to keep Curve2 noob friendly. Experienced people can do whatever they want with CSS.

@MustafaCan They are not cool dudes, and they don't have any flexibility (with table/tr/td not with display: table-cell/row etc...).

**Comment by Rupurudu** [comment]
You can type this if you want flexibility:

```
table, tbody, tr, td {
    display: block;
}
```

However I still believe tables are better for this job.

**Comment by live627** [comment]
+1,,  board index and message index are tabular data.

On Sun, Sep 6, 2015 at 1:51 AM, Mustafa Can Elmacı <notifications@github.com

**Comment by XinYenFon** [comment]
Sure go on rewrite the templates again I don't have any objection.

**Comment by illori** [comment]
it was in a table, but some team members thought it was not tabular data and did not need to stay in a table that is why the table got removed.

**Comment by StealthWombat** [comment]
Who said anything about rebuilding the template? All I'm talking about is a minor CSS change. It's basically as simple as changing float: left; to display: table-cell;

If you don't know what that means: https://developer.mozilla.org/en-US/docs/Web/CSS/display?redirectlocale=en-US&redirectslug=CSS%2Fdisplay

I'm not suggesting the template be reverted to the old table-based markup. That would be a step backwards. It's terrible semantics (no, board index content is not tabular as such) and it's useless for responsive.

As MDN says, what display: table-cell; does is make your row of divs behave like table cells. They will always stay in one row, even if the sum of their widths adds up to 150% of the available width. It won't matter, because they will behave like table cells. They will be rubbery enough to adjust themselves to fit. 

This has good support, so is not a problem for any browser you're likely to encounter these days.

This gives you the stability of the evil old table-based markup but without the terrible semantics, and allows you to easily override it with media queries later if you want a different display on smaller screens. You can change them to display: block; or inline-block or whatever you like.

In other words: doing it this way takes very little work and the result will be more n00b-friendly. That's why I'm suggesting it.

**Comment by StealthWombat** [comment]
"I like that idea.

Also, how would it work with blocks hon the stats page?"

It wouldn't suit the stats page because of the markup there. You'd need to wrap each row of two blocks in another div. If you just declared the lot of them as table-cell they'd all try to cram themselves into one row.

Floats are fine for the stats page, providing nobody adds horizontal padding or borders to the divs. if anyone does that, it will blow apart. You could make it safer by adding box-sizing: border-box; to those divs. That way any added borders or horizontal paddings would still be contained within an overall width of 50%, rather than being added on top of it, so the layout would still be stable. I think border-box is mass declared for some elements, but I'm not sure if it has been declared anywhere for the stats divs.

The other alternative would be to use inline-block, but that's a bit tricky on the stats page because the margins between the blocks aren't too predictable. They average around 0.3em, so can't be declared as a percentage of width; and it varies with font-family so is nasty if you're relying on a font stack.

You'd get automatic RTL support, but you wouldn't get a pixel perfect 50/50 split on width. You'd have to size them slightly under 50% width and use text-align: center; on the surrounding roundframe to get the blocks smack in the middle. Obviously this means you'd also have to declare text-align: left; on the content of the stats block divs to make the text sit properly.

Short version: inline-block is great for things like menus and button strips, but the stats page is probably better staying with floats on the current markup.

ETA: Oh and if you're wondering why floats were used for menus in 2.0.x when inline-block is better, that's because back then we had to support rotten old browsers that didn't know about inline-block. Not a problem these days. Any browser you're likely to run into will support it.

**Comment by StealthWombat** [comment]
@MustafaCan  "First, let's talk about why someone decided to ditch the original design. Why we are not using real tables anymore?"

Because the page isn't tabular. :) Tabular data is when you have one, and only one, item per cell. With the board index, most cells have several items. The first cell only contains one linked icon, but the second cell has board name, board description, list of moderators, and warning icons. The third cell has two items (topics and posts) and the fourth cell has date/time, last post title/link, plus last poster's name. To get all of that truly tabular you'd need ten cells in each row.

**Comment by Rupurudu** [comment]
@StealthWombat I don't understand. Isn't `<td>table cell</td>` and `<div style="display:table-cell">table cell</div>` the same thing? Or am I skipping something here?

**Comment by StealthWombat** [comment]
No, they're different. Best way to see the difference is to whip up some basic HTML on local and play around with them. You can read the docs on them too if you're really bored.

If you set a div (or anything else) to display: table-cell; it will behave like a table cell, but it's still a div. They invented this specifically for situations where you want this behaviour but don't want to use tables as such. It's really handy in the right places.

Say you wanted to do your responsive board index so that it was rock solid stable as a row of four divs, even if newbies did silly things with it, but on a 640 screen you wanted the .lastpost div content sitting beneath the .info div content.

So you have this for large screens:
`
.up_contain>div {display: table-cell;}
/* Then set your relative widths for each div, as usual*/
`
The .up_contain class will behave as a table row without being declared as such. Using the selectors this way just sets all the child divs at once.

On the 640 screen you can use this:
`
.info, .lastpost {display: block;}
`
to stack them down the page, which is something you can't do with real tables.

**Comment by StealthWombat** [comment]
Actually forget all that bollocks about using floats on the stats page (back here - https://github.com/SimpleMachines/SMF2.1/issues/2665#issuecomment-138553737).

I took a look at the current code and remembered why I had done them as inline-block. I'd forgotten about that. Short version is inline-block behaves better than floats when you have varying amount of content in the various stats blocks. Floats can get pretty hairy under those circumstances, so I used inline-block and fudged with margins and widths to get it stable.

Since it works, I'd stick with it. No point making extra work when it's not needed.

**Comment by XinYenFon** [comment]
@StealthWombat do we need anything more from here?

**Comment by StealthWombat** [comment]
Not as far as I can tell.

Don't quote me on that. :D

**Comment by XinYenFon** [comment]
Anyone has anything to add here?

**Comment by Rupurudu** [comment]
Nope.
